packages/recommend-react/src/Recommend.tsx

```diff
@@ -0,0 +1,235 @@
+        const cachedValue = cache[getCacheKey(widget.param.queries)];
+        widget.onResult(cachedValue);
+        return null; // No need to call API if we have value from cache
+      }
+      return widget.param;
+    })
+    .filter(isPresent);
+}
+
+const getQueryKeys = (params: GetParametersResult[]) => {
+  const queries = params
+    .map((p) => p.queries)
+    .reduce((a, b) => a.concat(b), []);
+  const keys = params.map((p) => p.keyPair);
+  return { queries, keys };
+};
+// Not sure about using "unknown"/"any" here ?
+const reducer: React.Reducer<StateType<unknown>, Action<any>> = (
+  state,
+  action
+) => {
+  const { type } = action;
+  switch (type) {
+    case 'register': {
+      const { widget } = action;
+      const param = widget.getParameters();
+      const isPreviouslyRegistered = isRegistered(state.widgets, param);
+      if (!isPreviouslyRegistered) {
+        return {
+          ...state,
+          widgets: [...state.widgets, { ...widget, param }],
+          isDirty: Date.now(),
+        };
+      }
+
+      return state;
+    }
+
+    case 'unregister': {
+      const { key } = action;
+      state.widgets = state.widgets.filter((w) => w.key !== key);
+      return state;
+    }
+
+    case 'request_success': {
+      return { ...state, status: 'idle', isDirty: null };
+    }
+
+    default:
+      return state;
+  }
+};
+
+export function Recommend<TObject>({
+  recommendClient: userRecommendClient,
+  children,
+}: RecommendProps) {
+  const recommendClient = useStableValue(userRecommendClient);
+
+  const [state, dispatch] = React.useReducer(reducer, {
+    isDirty: null,
+    cache: {},
+    widgets: [],
+    recommendClient,
+  });
+
+  React.useEffect(() => {
+    if (!state.isDirty) {
+      return;
+    }
+
+    const params = getCachedPrams<unknown>(state.widgets, state.cache); // ⚠️ Has side-effects
+    if (params.length === 0) {
+      return; // Return early if everything is cached
+    }
+    const { queries, keys } = getQueryKeys(params);
+
+    keys.forEach((key) => {
+      state.widgets.forEach((widget) => {
+        if (widget.key === key.key) {
+          widget.onRequest();
+        }
+      });
+    });
+
+    getBatchRecommendations<TObject>({
+      recommendClient,
+      queries,
+      keys,
+    }).then((result) => {
+      Object.entries(result).forEach(([key, value]) => {
+        state.widgets.forEach((widget) => {
+          if (widget.key === key) {
+            widget.onResult(value);
+            state.cache[getCacheKey(widget.param.queries)] = value;
+          }
+        });
+      });
+
+      dispatch({ type: 'request_success' });
+    });
+  }, [state.isDirty, dispatch, state.widgets, state.cache, recommendClient]);
+
+  const register = React.useCallback(
+    (widget: Omit<RecommendWidget<TObject>, 'param'>) => {
+      dispatch({ type: 'register', widget });
+      return (key: string) => {
+        dispatch({ type: 'unregister', key });
+      };
```

Comment:
At this point we're already aware of the `key` via `widget.key` so we don't need to pass `key` to unregister.

Reply:
Ah yes ! It's an artifact of a previous approach 👍 Good catch